File: go/logger.go
package cordis

import (
	"encoding/json"
	"errors"
	"fmt"
	"io"
	"os"
	"strconv"
	"strings"
	"sync"
	"time"
)

// Level is a log severity. Lower is more severe, mirroring LoggerLevel
// upstream: a message is exported when the target level is greater than or
// equal to the message level.
type Level int

const (
	LevelError Level = 0
	LevelWarn  Level = 1
	LevelInfo  Level = 2
	LevelDebug Level = 3
	// LevelUnset marks an absent level override.
	LevelUnset Level = -1
)

func (l Level) String() string {
	switch l {
	case LevelError:
		return "error"
	case LevelWarn:
		return "warn"
	case LevelInfo:
		return "info"
	case LevelDebug:
		return "debug"
	}
	return "unknown"
}

// Message is one log record, mirroring the Message interface upstream.
type Message struct {
	SN    int64
	Time  time.Time
	Name  string
	Type  string
	Level Level
	Args  []any
}

// Exporter receives log messages. Implementations must be safe for
// concurrent use.
type Exporter interface {
	Export(Message)
}

// ExporterFunc adapts a function to Exporter.
type ExporterFunc func(Message)

func (f ExporterFunc) Export(m Message) { f(m) }

// LoggerIntercept overrides logger name and level for a scope, mirroring the
// logger intercept upstream. Install it with Context.Intercept:
//
//	ctx = ctx.Intercept("logger", cordis.LoggerIntercept{Name: "worker"})
type LoggerIntercept struct {
	Name  string
	Level Level
}

type exporterEntry struct {
	exporter Exporter
	levels   map[string]Level
}

// loggerService is the per-tree logging facility, mirroring LoggerService
// upstream. It fans messages out to registered exporters and keeps a bounded
// in-memory buffer of the most recent messages.
type loggerService struct {
	mu         sync.Mutex
	snMessage  int64
	snExporter int
	exporters  map[int]*exporterEntry

	bufferMu   sync.Mutex
	bufferSize int
	buffer     []Message
}

func newLoggerService() *loggerService {
	s := &loggerService{
		exporters:  make(map[int]*exporterEntry),
		bufferSize: 1000,
	}
	s.AddExporter(ExporterFunc(s.bufferExport))
	return s
}

// bufferExport is the default exporter keeping the last bufferSize messages.
func (s *loggerService) bufferExport(m Message) {
	s.bufferMu.Lock()
	defer s.bufferMu.Unlock()
	s.buffer = append(s.buffer, m)
	if overflow := len(s.buffer) - s.bufferSize; overflow > 0 {
		s.buffer = append([]Message(nil), s.buffer[overflow:]...)
	}
}

// Buffer returns a copy of the most recent messages, oldest first.
func (s *loggerService) Buffer() []Message {
	s.bufferMu.Lock()
	defer s.bufferMu.Unlock()
	return append([]Message(nil), s.buffer...)
}

// SetBufferSize resizes the message buffer, truncating the oldest entries.
func (s *loggerService) SetBufferSize(size int) {
	s.bufferMu.Lock()
	defer s.bufferMu.Unlock()
	s.bufferSize = size
	if overflow := len(s.buffer) - size; overflow > 0 {
		s.buffer = append([]Message(nil), s.buffer[overflow:]...)
	}
}

// AddExporter registers an exporter and returns a Disposer removing it.
// levels maps logger names to their minimum exported level; the "default"
// key applies to every other name. Nil levels export everything.
func (s *loggerService) AddExporter(e Exporter, levels ...map[string]Level) Disposer {
	s.mu.Lock()
	defer s.mu.Unlock()
	s.snExporter++
	id := s.snExporter
	entry := &exporterEntry{exporter: e}
	if len(levels) > 0 {
		entry.levels = levels[0]
	}
	s.exporters[id] = entry
	return func() {
		s.mu.Lock()
		defer s.mu.Unlock()
		delete(s.exporters, id)
	}
}

// ClearExporters removes every exporter, including the default buffer.
func (s *loggerService) ClearExporters() {
	s.mu.Lock()
	defer s.mu.Unlock()
	s.exporters = make(map[int]*exporterEntry)
}

func (s *loggerService) dispatch(name string, level Level, args []any) {
	s.mu.Lock()
	s.snMessage++
	sn := s.snMessage
	entries := make([]*exporterEntry, 0, len(s.exporters))
	for _, entry := range s.exporters {
		target := LevelInfo
		if entry.levels != nil {
			if l, ok := entry.levels[name]; ok {
				target = l
			} else if l, ok := entry.levels["default"]; ok {
				target = l
			}
		}
		if target < level {
			continue
		}
		entries = append(entries, entry)
	}
	s.mu.Unlock()

	if len(entries) == 0 {
		return
	}
	msg := Message{
		SN:    sn,
		Time:  time.Now(),
		Name:  name,
		Type:  level.String(),
		Level: level,
		Args:  args,
	}
	for _, entry := range entries {
		entry.exporter.Export(msg)
	}
}

// logError is the framework's internal error channel, mirroring
// ctx.logger.error upstream.
func (c *core) logError(name string, err error) {
	if name == "" {
		name = "root"
	}
	c.logger.dispatch(name, LevelError, []any{err})
}

// Logger is a named, leveled logging handle, mirroring Logger upstream.
type Logger struct {
	service *loggerService
	name    string
	level   Level
}

// Logger returns a logger for this context. Name resolution order mirrors
// upstream: an explicit argument wins, then the nearest logger intercept,
// then the name of the fiber owning the context.
func (c *Context) Logger(names ...string) *Logger {
	l := &Logger{service: c.core.logger, level: LevelUnset}
	if len(names) > 0 && names[0] != "" {
		l.name = names[0]
	}
	if intercept, ok := c.Intercepted("logger"); ok {
		if li, ok := intercept.(LoggerIntercept); ok {
			if l.name == "" {
				l.name = li.Name
			}
			l.level = li.Level
		}
	}
	if l.name == "" {
		l.name = c.fiber.Name()
	}
	return l
}

// Name returns the logger name.
func (l *Logger) Name() string { return l.name }

// Error logs at LevelError. A single error argument is formatted with its
// message; wrapped and joined errors are expanded like upstream.
func (l *Logger) Error(args ...any) { l.log(LevelError, args) }

// Warn logs at LevelWarn.
func (l *Logger) Warn(args ...any) { l.log(LevelWarn, args) }

// Info logs at LevelInfo.
func (l *Logger) Info(args ...any) { l.log(LevelInfo, args) }

// Debug logs at LevelDebug.
func (l *Logger) Debug(args ...any) { l.log(LevelDebug, args) }

func (l *Logger) log(level Level, args []any) {
	// Expand a lone error argument the way upstream does: joined errors fan
	// out, wrapped errors unwrap.
	if len(args) == 1 {
		if err, ok := args[0].(error); ok {
			if joined, ok := err.(interface{ Unwrap() []error }); ok {
				for _, inner := range joined.Unwrap() {
					l.log(level, []any{inner})
				}
				return
			}
			if inner := errors.Unwrap(err); inner != nil {
				l.log(level, []any{inner})
				return
			}
		}
	}
	if l.level != LevelUnset && l.level < level {
		return
	}
	l.service.dispatch(l.name, level, args)
}

// FormatMessage renders a message to a single string, mirroring the format
// pipeline upstream: the first argument is a printf style format string
// supporting %s, %d, %i, %f, %o, %O and %% verbs, and remaining arguments
// are appended.
func FormatMessage(m Message) string {
	args := append([]any(nil), m.Args...)
	if len(args) > 0 {
		if err, ok := args[0].(error); ok {
			args[0] = err.Error()
			args = append([]any{"%s"}, args...)
		} else if _, ok := args[0].(string); !ok {
			args = append([]any{"%o"}, args...)
		}
	}
	if len(args) == 0 {
		return ""
	}

	format, _ := args[0].(string)
	args = args[1:]

	var b strings.Builder
	argIndex := 0
	for i := 0; i < len(format); i++ {
		if format[i] != '%' || i+1 >= len(format) {
			b.WriteByte(format[i])
			continue
		}
		verb := format[i+1]
		i++
		if verb == '%' {
			b.WriteByte('%')
			continue
		}
		if argIndex >= len(args) {
			b.WriteByte('%')
			b.WriteByte(verb)
			continue
		}
		value := args[argIndex]
		argIndex++
		switch verb {
		case 's':
			fmt.Fprint(&b, value)
		case 'd', 'i':
			fmt.Fprint(&b, truncateNumber(value))
		case 'f':
			fmt.Fprint(&b, toFloat(value))
		case 'o', 'O':
			b.WriteString(toJSON(value))
		case 'c', 'C':
			// Color directives render as plain text in Go.
		default:
			b.WriteByte('%')
			b.WriteByte(verb)
		}
	}
	for ; argIndex < len(args); argIndex++ {
		b.WriteByte(' ')
		if err, ok := args[argIndex].(error); ok {
			b.WriteString(err.Error())
		} else {
			fmt.Fprint(&b, args[argIndex])
		}
	}
	return b.String()
}

func truncateNumber(value any) string {
	switch v := value.(type) {
	case int:
		return strconv.Itoa(v)
	case int64:
		return strconv.FormatInt(v, 10)
	case float64:
		return strconv.FormatInt(int64(v), 10)
	case float32:
		return strconv.FormatInt(int64(v), 10)
	case string:
		if f, err := strconv.ParseFloat(v, 64); err == nil {
			return strconv.FormatInt(int64(f), 10)
		}
		return "0"
	}
	return fmt.Sprint(value)
}

func toFloat(value any) string {
	switch v := value.(type) {
	case float64:
		return strconv.FormatFloat(v, 'g', -1, 64)
	case float32:
		return strconv.FormatFloat(float64(v), 'g', -1, 32)
	case int:
		return strconv.Itoa(v)
	case int64:
		return strconv.FormatInt(v, 10)
	}
	return fmt.Sprint(value)
}

func toJSON(value any) string {
	data, err := json.Marshal(value)
	if err != nil {
		return fmt.Sprint(value)
	}
	return string(data)
}

// ConsoleExporter writes formatted messages to an io.Writer, mirroring the
// logger-console package upstream.
type ConsoleExporter struct {
	W io.Writer
}

// NewConsoleExporter returns an exporter writing to w, or os.Stderr when w
// is nil.
func NewConsoleExporter(w io.Writer) *ConsoleExporter {
	if w == nil {
		w = os.Stderr
	}
	return &ConsoleExporter{W: w}
}

// Export implements Exporter.
func (e *ConsoleExporter) Export(m Message) {
	fmt.Fprintf(e.W, "[%s] %s: %s\n", m.Type, m.Name, FormatMessage(m))
}

File: rust/src/logger.rs
//! The logger service: levels, exporters and a bounded message buffer.
//!
//! Mirrors `LoggerService` upstream (and the Go port's `logger.go`): every
//! context tree owns one service that fans messages out to registered
//! exporters and keeps the most recent messages in memory. The framework's
//! own error channel (failing plugin bodies, panicking cleanups) dispatches
//! into this service at [`Level::Error`]; [`crate::Context::logged_errors`]
//! reads those entries back.
//!
//! Exporters run with no framework borrows held, so they may freely call
//! back into the tree.

use std::collections::HashMap;
use std::time::SystemTime;

use crate::context::{Context, Disposer};
use crate::core::Core;
use crate::sync::{Rc, RefCell};
#[cfg(feature = "thread-safe")]
use crate::sync::BorrowExt as _;

/// A log severity. The variant order is the severity order, mirroring
/// `LoggerLevel` upstream: a message is exported when the target level is
/// greater than or equal to the message level.
#[derive(Debug, Clone, Copy, PartialEq, Eq, PartialOrd, Ord, Hash)]
pub enum Level {
    /// Failures: a plugin body or cleanup failed.
    Error,
    /// Recoverable problems.
    Warn,
    /// Ordinary operational output.
    Info,
    /// Verbose diagnostics.
    Debug,
}

impl Level {
    /// The canonical name, used as the message kind.
    #[must_use]
    pub const fn as_str(self) -> &'static str {
        match self {
            Self::Error => "error",
            Self::Warn => "warn",
            Self::Info => "info",
            Self::Debug => "debug",
        }
    }
}

/// One typed log argument.
///
/// Rust has no `any` with reflection; arguments
/// carry their rendering class instead, and [`format_message`] applies the
/// printf verbs to it. [`Arg::Json`] is the `%o`/`%O` channel: pass
/// pre-serialized JSON (the crate is dependency free and does not serialize).
#[derive(Debug, Clone, PartialEq)]
pub enum Arg {
    /// Rendered with `%s` and plain display.
    Str(String),
    /// An integer, rendered identically by `%s`, `%d` and `%i`.
    Int(i64),
    /// A float; `%d` truncates it like upstream.
    Float(f64),
    /// An error message; a lone error argument formats through `%s`.
    Error(String),
    /// Pre-serialized JSON, rendered verbatim by `%o`/`%O`.
    Json(String),
}

impl std::fmt::Display for Arg {
    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
        match self {
            Self::Str(s) | Self::Error(s) | Self::Json(s) => write!(f, "{s}"),
            Self::Int(i) => write!(f, "{i}"),
            Self::Float(x) => write!(f, "{x}"),
        }
    }
}

impl From<&str> for Arg {
    fn from(v: &str) -> Self {
        Self::Str(v.to_string())
    }
}

impl From<String> for Arg {
    fn from(v: String) -> Self {
        Self::Str(v)
    }
}

impl From<i64> for Arg {
    fn from(v: i64) -> Self {
        Self::Int(v)
    }
}

impl From<f64> for Arg {
    fn from(v: f64) -> Self {
        Self::Float(v)
    }
}

impl From<crate::Error> for Arg {
    fn from(v: crate::Error) -> Self {
        Self::Error(v.to_string())
    }
}

/// One log record, mirroring the `Message` interface upstream.
#[derive(Debug, Clone, PartialEq)]
pub struct Message {
    /// The tree-wide sequence number.
    pub sn: i64,
    /// When the message was dispatched.
    pub time: SystemTime,
    /// The logger name.
    pub name: String,
    /// The severity name, [`Level::as_str`].
    pub kind: &'static str,
    /// The severity.
    pub level: Level,
    /// The formatted arguments.
    pub args: Vec<Arg>,
}

/// Receives log messages. Implementations must be safe for concurrent use
/// under the `thread-safe` feature.
pub trait Exporter: crate::sync::Shared {
    /// Deliver one message.
    fn export(&self, message: &Message);
}

/// The logger intercept value, installed with [`Context::intercept`]:
///
/// ```
/// use cordis::{Context, Level, LoggerIntercept};
///
/// let ctx = Context::new();
/// let scoped = ctx.intercept("logger", cordis::value(LoggerIntercept {
///     name: Some("worker".to_string()),
///     level: Some(Level::Warn),
/// }));
/// assert_eq!(scoped.logger(None).name(), "worker");
/// ```
#[derive(Debug, Clone, Default, PartialEq, Eq)]
pub struct LoggerIntercept {
    /// The logger name override.
    pub name: Option<String>,
    /// The level override: messages above it are dropped before dispatch.
    pub level: Option<Level>,
}

enum Sink {
    /// The built-in bounded buffer.
    Buffer,
    /// A user exporter.
    Exporter(Rc<dyn Exporter>),
}

struct ExporterEntry {
    sink: Sink,
    levels: Option<HashMap<String, Level>>,
}

/// The per-tree logging facility. Owned by the core; reach it through
/// [`Context::logger`], [`Context::add_exporter`] and friends.
#[derive(Default)]
pub struct LoggerService {
    sn_message: i64,
    sn_exporter: i64,
    exporters: Vec<(i64, ExporterEntry)>,
    buffer: Vec<Message>,
    buffer_size: usize,
}

impl LoggerService {
    /// A service with the default buffer exporter (1000 messages) attached.
    pub fn new() -> Self {
        let mut service = Self {
            buffer_size: 1000,
            ..Self::default()
        };
        service.sn_exporter = service.sn_exporter.wrapping_add(1);
        service.exporters.push((
            service.sn_exporter,
            ExporterEntry {
                sink: Sink::Buffer,
                levels: None,
            },
        ));
        service
    }

    /// The most recent messages, oldest first.
    #[must_use]
    pub fn buffer(&self) -> &[Message] {
        &self.buffer
    }

    /// Resize the message buffer, truncating the oldest entries.
    pub fn set_buffer_size(&mut self, size: usize) {
        self.buffer_size = size;
        let overflow = self.buffer.len().saturating_sub(size);
        if overflow > 0 {
            self.buffer.drain(..overflow);
        }
    }

    /// Register an exporter and return a Disposer removing it. `levels`
    /// maps logger names to their minimum exported level; the `"default"`
    /// key applies to every other name. No map means `info`.
    pub fn add_exporter(
        &mut self,
        exporter: Rc<dyn Exporter>,
        levels: Option<HashMap<String, Level>>,
    ) -> i64 {
        self.sn_exporter = self.sn_exporter.wrapping_add(1);
        let id = self.sn_exporter;
        self.exporters.push((
            id,
            ExporterEntry {
                sink: Sink::Exporter(exporter),
                levels,
            },
        ));
        id
    }

    /// Remove every exporter, including the built-in buffer.
    pub fn clear_exporters(&mut self) {
        self.exporters.clear();
    }

    fn remove_exporter(&mut self, id: i64) {
        self.exporters.retain(|(candidate, _)| *candidate != id);
    }

    /// Assign the sequence number, append to the buffer and select the
    /// exporters whose level target admits the message. The callers run the
    /// selected exporters after every framework borrow is released.
    fn prepare(&mut self, name: &str, level: Level, args: Vec<Arg>) -> (Message, Vec<Rc<dyn Exporter>>) {
        self.sn_message = self.sn_message.wrapping_add(1);
        let message = Message {
            sn: self.sn_message,
            time: SystemTime::now(),
            name: name.to_string(),
            kind: level.as_str(),
            level,
            args,
        };
        if self.buffer.len() >= self.buffer_size && self.buffer_size > 0 {
            self.buffer.remove(0);
        }
        if self.buffer_size > 0 {
            self.buffer.push(message.clone());
        }
        let exporters = self
            .exporters
            .iter()
            .filter_map(|(_, entry)| {
                let target = entry.levels.as_ref().map_or(Level::Info, |levels| {
                    levels.get(name).or_else(|| levels.get("default")).copied().unwrap_or(Level::Info)
                });
                if target < level {
                    return None;
                }
                match &entry.sink {
                    Sink::Exporter(exporter) => Some(Rc::clone(exporter)),
                    Sink::Buffer => None,
                }
            })
            .collect();
        (message, exporters)
    }
}

/// A named, leveled logging handle, mirroring `Logger` upstream. Create one
/// with [`Context::logger`].
pub struct Logger {
    core: Rc<RefCell<Core>>,
    name: String,
    level: Option<Level>,
}

impl Logger {
    /// The logger name.
    #[must_use]
    pub fn name(&self) -> &str {
        &self.name
    }

    /// Log at [`Level::Error`].
    pub fn error(&self, args: &[Arg]) {
        self.log(Level::Error, args);
    }

    /// Log at [`Level::Warn`].
    pub fn warn(&self, args: &[Arg]) {
        self.log(Level::Warn, args);
    }

    /// Log at [`Level::Info`].
    pub fn info(&self, args: &[Arg]) {
        self.log(Level::Info, args);
    }

    /// Log at [`Level::Debug`].
    pub fn debug(&self, args: &[Arg]) {
        self.log(Level::Debug, args);
    }

    /// Dispatch at `level` unless the handle's level intercept rejects it.
    pub fn log(&self, level: Level, args: &[Arg]) {
        if self.level.is_some_and(|l| l < level) {
            return;
        }
        dispatch(&self.core, &self.name, level, args.to_vec());
    }
}

/// Dispatch one message: prepare under the core borrow (sequence number,
/// buffer, exporter selection), then run the user exporters with no borrow
/// held.
fn dispatch(core: &Rc<RefCell<Core>>, name: &str, level: Level, args: Vec<Arg>) {
    let (message, exporters) = {
        let mut c = core.borrow_mut();
        c.logger.prepare(name, level, args)
    };
    for exporter in &exporters {
        exporter.export(&message);
    }
}

/// The framework's internal error channel, mirroring `ctx.logger.error`
/// upstream. A missing name falls back to `"root"`.
pub fn log_error(core: &Rc<RefCell<Core>>, name: &str, message: &str) {
    let name = if name.is_empty() { "root" } else { name };
    dispatch(core, name, Level::Error, vec![Arg::Error(message.to_string())]);
}

/// Render a message to a single string, mirroring the format pipeline
/// upstream.
///
/// The first argument is a printf style format string supporting
/// `%s`, `%d`, `%i`, `%f`, `%o`, `%O` and `%%` verbs, and remaining
/// arguments are appended separated by spaces. A lone error argument
/// formats through `%s`; any other non-string first argument through `%o`.
#[must_use]
pub fn format_message(message: &Message) -> String {
    let mut args = message.args.clone();
    let prepend = match args.first() {
        Some(Arg::Error(text)) => {
            let text = text.clone();
            if let Some(first) = args.first_mut() {
                *first = Arg::Str(text);
            }
            Some("%s")
        }
        Some(Arg::Str(_)) => None,
        Some(_) => Some("%o"),
        None => return String::new(),
    };
    if let Some(verb) = prepend {
        args.insert(0, Arg::Str(verb.to_string()));
    }
    let Some((Arg::Str(format), rest)) = args.split_first() else {
        return String::new();
    };

    let mut out = String::new();
    let mut arg_index = 0usize;
    let mut chars = format.chars();
    while let Some(c) = chars.next() {
        if c != '%' {
            out.push(c);
            continue;
        }
        let Some(verb) = chars.next() else {
            out.push(c);
            continue;
        };
        if verb == '%' {
            out.push('%');
            continue;
        }
        let Some(value) = rest.get(arg_index) else {
            out.push('%');
            out.push(verb);
            continue;
        };
        arg_index = arg_index.saturating_add(1);
        match verb {
            's' => out.push_str(&value.to_string()),
            'd' | 'i' => out.push_str(&truncate_number(value)),
            'f' => out.push_str(&render_float(value)),
            'o' | 'O' => match value {
                Arg::Json(json) => out.push_str(json),
                other => out.push_str(&other.to_string()),
            },
            'c' | 'C' => {}
            _ => {
                out.push('%');
                out.push(verb);
            }
        }
    }
    for value in rest.iter().skip(arg_index) {
        out.push(' ');
        out.push_str(&value.to_string());
    }
    out
}

/// `%d`/`%i`: integers pass through, floats truncate, numeric strings
/// parse then truncate, anything else renders as `0` like upstream. The
/// float truncation is the upstream behavior, hence the deliberate casts.
#[allow(clippy::cast_possible_truncation, clippy::as_conversions)]
fn truncate_number(value: &Arg) -> String {
    match value {
        Arg::Int(i) => i.to_string(),
        Arg::Float(x) => format!("{}", *x as i64),
        Arg::Str(s) => {
            let parsed: f64 = s.parse().unwrap_or(0.0);
            format!("{}", parsed as i64)
        }
        Arg::Error(text) | Arg::Json(text) => {
            let parsed: f64 = text.parse().unwrap_or(0.0);
            format!("{}", parsed as i64)
        }
    }
}

/// `%f`: floats render losslessly, integers as themselves.
fn render_float(value: &Arg) -> String {
    match value {
        Arg::Float(x) => format!("{x}"),
        Arg::Int(i) => i.to_string(),
        Arg::Str(s) | Arg::Error(s) | Arg::Json(s) => s.clone(),
    }
}

/// An exporter writing formatted messages to a [`std::io::Write`] sink,
/// mirroring the `logger-console` package upstream.
///
/// Writes are best effort:
/// a logger cannot report its own write failure through logging.
pub struct ConsoleExporter<W> {
    out: RefCell<W>,
}

impl<W: std::io::Write + crate::sync::Shared> ConsoleExporter<W> {
    /// Wrap a writer. [`ConsoleExporter::stderr`] is the common choice.
    #[must_use]
    pub const fn new(out: W) -> Self {
        Self { out: RefCell::new(out) }
    }
}

impl ConsoleExporter<std::io::Stderr> {
    /// The standard error console exporter.
    #[must_use]
    pub fn stderr() -> Self {
        Self::new(std::io::stderr())
    }
}

impl<W: std::io::Write + crate::sync::Shared> Exporter for ConsoleExporter<W> {
    fn export(&self, message: &Message) {
        let line = format!("[{}] {}: {}\n", message.kind, message.name, format_message(message));
        // Deliberate discard: a logger cannot report its own write failure
        // through logging.
        let _ = self.out.borrow_mut().write_all(line.as_bytes());
    }
}

impl Context {
    /// A logger for this context. Name resolution mirrors upstream: an
    /// explicit non-empty argument wins, then the nearest
    /// [`LoggerIntercept`] installed through [`Context::intercept`], then
    /// the name of the fiber owning the context.
    #[must_use]
    pub fn logger(&self, name: Option<&str>) -> Logger {
        let resolved = name.filter(|n| !n.is_empty()).map(std::string::ToString::to_string);
        let (resolved, level) = if let Some(intercept) = self.intercepted("logger")
            && let Some(li) = intercept.downcast_ref::<LoggerIntercept>()
        {
            (resolved.or_else(|| li.name.clone()), li.level)
        } else {
            (resolved, None)
        };
        Logger {
            core: Rc::clone(&self.core),
            name: resolved.unwrap_or_else(|| self.fiber().name()),
            level,
        }
    }

    /// Register a log exporter on this tree and return a Disposer removing
    /// it. `levels` maps logger names to their minimum exported level; the
    /// `"default"` key applies to every other name. `None` exports
    /// everything at `info` or above.
    pub fn add_exporter(
        &self,
        exporter: Rc<dyn Exporter>,
        levels: Option<HashMap<String, Level>>,
    ) -> Disposer {
        let id = {
            let mut core = self.core.borrow_mut();
            core.logger.add_exporter(exporter, levels)
        };
        Disposer::new({
            let core = Rc::clone(&self.core);
            move || {
                core.borrow_mut().logger.remove_exporter(id);
            }
        })
    }

    /// Remove every log exporter, including the built-in buffer.
    pub fn clear_exporters(&self) {
        self.core.borrow_mut().logger.clear_exporters();
    }

    /// A copy of the most recent log messages, oldest first.
    #[must_use]
    pub fn logger_buffer(&self) -> Vec<Message> {
        self.core.borrow().logger.buffer().to_vec()
    }

    /// Resize the message buffer, truncating the oldest entries.
    pub fn set_logger_buffer_size(&self, size: usize) {
        self.core.borrow_mut().logger.set_buffer_size(size);
    }
}
